src/project/config.rs

```diff
@@ -0,0 +1,302 @@
+                enum Field {
+                    $(
+                        $field,
+                    )*
+                }
+
+                impl ::serde::de::Deserialize for Field {
+                    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+                        where D: ::serde::de::Deserializer
+                    {
+                        struct FieldVisitor;
+
+                        impl ::serde::de::Visitor for FieldVisitor {
+                            type Value = Field;
+
+                            fn expecting(&self, formatter: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+                                try!(write!(formatter, "one of "));
+
+                                $(
+                                    try!(write!(formatter, "`{}`, ", stringify!($field)));
+                                )*
+
+                                Ok(())
+                            }
+
+                            fn visit_str<E>(self, value: &str) -> Result<Field, E>
+                                where E: ::serde::de::Error,
+                            {
+                                match value {
+                                    $(
+                                        stringify!($field) => Ok(Field::$field),
+                                    )*
+                                    _ => {
+                                        Err(::serde::de::Error::unknown_field(value, FIELDS))
+                                    },
+                                }
+                            }
+                        }
+
+                        deserializer.deserialize_struct_field(FieldVisitor)
+                    }
+                }
+
+                struct IdentifierVisitor;
+                impl ::serde::de::Visitor for IdentifierVisitor {
+                    type Value = $identifier;
+
+                    fn expecting(&self, formatter: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+                        write!(formatter, "a {} configuration", stringify!($identifier))
+                    }
+
+                    fn visit_seq<V>(self, mut visitor: V) -> Result<Self::Value, V::Error>
+                        where V: ::serde::de::SeqVisitor,
+                    {
+                        let result = $identifier {
+                            $(
+                                $field: {
+                                    match visitor.visit() {
+                                        Ok(opt) => match opt {
+                                            Some(v) => v,
+                                            None => {
+                                                $default
+                                            },
+                                        },
+                                        Err(err) => {
+                                            println!("{}: {}, expecting: `{}`", CONFIG_NAME, err, stringify!($ty));

```

Comment:
I was waiting for the logging stuff with `slog` to be implemented, not sure what to do until then.